opt: support SHOW RANGES + TableIndexName cleanup

RaduBerinde opened the pull request:
#### sql: use TableIndexName for the case of a table and no index

Allow `TableIndexName` to have an empty index and explicit table (in
which case it refers to the primary index of that table). The zone
specifier was already using it in that way. This simplifies some of
the statements that have both TABLE and INDEX versions.

Release note: None

#### sql: move resolution of a TableIndexName to the catalog

Rewriting the code to resolve a TableIndexName in the catalog, and
rerouting the exiting code path.

Release note: None

#### opt: disable memo reuse for delegated statements

The delegate implementations don't register their dependencies with
the metadata. For some it would be very difficult (e.g. some search
for an index name through all tables in a catalog).

This change disables memo caching if we encountered any delegated
statements. We should probably do this for `CreateTable` as well, and
simplify the cache invalidation code to not care about schemas.

Release note: None

#### opt: support SHOW RANGES

Move the ShowRanges implementation to `delegate`.

Release note: None

RaduBerinde committed:
sql: use TableIndexName for the case of a table and no index

Allow `TableIndexName` to have an empty index and explicit table (in
which case it refers to the primary index of that table). The zone
specifier was already using it in that way. This simplifies some of
the statements that have both TABLE and INDEX versions.

Release note: None

RaduBerinde committed:
sql: move resolution of a TableIndexName to the catalog

Rewriting the code to resolve a TableIndexName in the catalog, and
rerouting the existing code path.

Release note: None

RaduBerinde committed:
opt: support SHOW RANGES

Move the ShowRanges implementation to `delegate`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37194)
<!-- Reviewable:end -->

RaduBerinde committed:
opt: disable memo reuse for delegated statements

The delegate implementations don't register their dependencies with
the metadata. For some it would be very difficult (e.g. some search
for an index name through all tables in a catalog).

This change disables memo caching if we encountered any delegated
statements. We should probably do this for `CreateTable` as well, and
simplify the cache invalidation code to not care about schemas.

Release note: None

rytaft reviewed (changes requested):
Nice improvements!

[nit] in the second commit message, change exiting to existing.

Also, it feels like the `ResolveTableIndex` code could use some additional tests.

> Reviewed 11 of 11 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 7 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-:-Ldu8ZT58O5Gsk1Rl7UA:b-vjlbk5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @justinj, @knz, and @RaduBerinde)

___
*[pkg/sql/opt_catalog.go, line 962 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-LduBKy1AZYYZF5uhGmt:-LduBKy1AZYYZF5uhGmu:b-s0qnxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb7a5b98e7e07a2155bb294f18bc7274334092c6/pkg/sql/opt_catalog.go#L962)):*
> ```Go
> func (oi *optIndex) Span() roachpb.Span {
> 	desc := oi.tab.desc
> 	if desc.ID < keys.MaxSystemConfigDescID {
> ```

A comment would help here to explain the purpose of this check
___
*[pkg/sql/resolver.go, line 544 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-Ldu8yL4BBHeUJKwdD5J:-Ldu8yL4BBHeUJKwdD5K:b-pi4zwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf19bcf126c84d881b732a398e519005460cfdf/pkg/sql/resolver.go#L544)):*
> ```Go
> //	}
> //	return tableDesc, index, nil
> //}
> ```

this is dead code now
___
*[pkg/sql/delegate/show_ranges.go, line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-LduBZ8B5XWOzyFa0fgd:-LduBZ8B5XWOzyFa0fge:btnj08c) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb7a5b98e7e07a2155bb294f18bc7274334092c6/pkg/sql/delegate/show_ranges.go#L1)):*
> ```Go
> // Copyright 2015 The Cockroach Authors.
> ```

[nit] 2019
___
*[pkg/sql/opt/cat/utils.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-LduAXTC5IUMS6i5Eq9s:-LduAXTD7kOlnItYtkac:bc10abq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf19bcf126c84d881b732a398e519005460cfdf/pkg/sql/opt/cat/utils.go#L105)):*
> ```Go
> 				if found != nil {
> 					return nil, pgerror.Newf(pgerror.CodeAmbiguousParameterError,
> 						"index name %q is ambiguous (found in %s and %s)",
> ```

is this tested anywhere?
___
*[pkg/sql/sem/tree/table_name.go, line 211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-Ldu89Pd4V6spplzHEwL:-Ldu89Pe4e5Qc7Uuk93d:b-qvfatt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c82beb8ef27106697ecd2cdf45ddfafed5f23f26/pkg/sql/sem/tree/table_name.go#L211)):*
> ```Go
> // Note that it is possible to specify the schema or catalog without specifying
> // a table name; in this case, Table.TableNamePrefix has the fields set but
> // Table.TableName is empty.
> ```

I'd merge this last section into your third bullet point above.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: tests for catalog utils

Add tests for `ExpandDataSourceGlob` and `ResolveTableIndex`.

Release note: None

RaduBerinde reviewed (commented):
TFTR!

Added unit tests for `ExpandDataSourceGlob` and `ResolveTableIndex`. They were tested indirectly through logictests, but it's good to have these too.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-:-LdzIPyQD7TG47OrCHHs:b-ke71uu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @justinj, @knz, and @rytaft)

___
*[pkg/sql/opt_catalog.go, line 962 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-LduBKy1AZYYZF5uhGmt:-LdzH5Uk454KB22W1y9S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb7a5b98e7e07a2155bb294f18bc7274334092c6/pkg/sql/opt_catalog.go#L962)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

A comment would help here to explain the purpose of this check
</blockquote></details>

Done.
___
*[pkg/sql/resolver.go, line 544 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-Ldu8yL4BBHeUJKwdD5J:-LdzHNjSEUJaYq1hMwVg:bb3lzi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf19bcf126c84d881b732a398e519005460cfdf/pkg/sql/resolver.go#L544)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

this is dead code now
</blockquote></details>

Oops, cleaned up.
___
*[pkg/sql/delegate/show_ranges.go, line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-LduBZ8B5XWOzyFa0fgd:-LdzHZZsB8QqV506iaTm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb7a5b98e7e07a2155bb294f18bc7274334092c6/pkg/sql/delegate/show_ranges.go#L1)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] 2019
</blockquote></details>

Done.
___
*[pkg/sql/opt/cat/utils.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-LduAXTC5IUMS6i5Eq9s:-LdzH_duE2bRlFAA2JcU:bko6iwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bf19bcf126c84d881b732a398e519005460cfdf/pkg/sql/opt/cat/utils.go#L105)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

is this tested anywhere?
</blockquote></details>

There are some logic tests (in drop_index). Added a unit test too.
___
*[pkg/sql/sem/tree/table_name.go, line 211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-Ldu89Pd4V6spplzHEwL:-LdzI5ocB9l5n6XLT_r_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c82beb8ef27106697ecd2cdf45ddfafed5f23f26/pkg/sql/sem/tree/table_name.go#L211)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I'd merge this last section into your third bullet point above.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 12 files at r5, 1 of 2 files at r6, 3 of 4 files at r7, 2 of 2 files at r8, 7 of 7 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-:-LeCOQkREM5SIwHkE8NM:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @justinj, @knz, and @RaduBerinde)

___
*[pkg/sql/opt/cat/utils_test.go, line 167 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-LeCO8IV29zxyG2lKPeB:-LeCO8IV29zxyG2lKPeC:b-oms7ox) ([raw file](https://github.com/cockroachdb/cockroach/blob/12f943d1f0e114fc184e7e27cc9e369c8fd2e653/pkg/sql/opt/cat/utils_test.go#L167)):*
> ```Go
> 			},
> 			expected: `error: index name "idx2" is ambiguous (found in t.public.c and t.public.b)`,
> 		},
> ```

Do you have any tests where the schema is provided (e.g. `public`) instead of the database name?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37194)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @justinj, @knz, and @rytaft)

___
*[pkg/sql/opt/cat/utils_test.go, line 167 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/37194#-LeCO8IV29zxyG2lKPeB:-LeDYmZFCDGV7Ywc1p5j:b3kcgww) ([raw file](https://github.com/cockroachdb/cockroach/blob/12f943d1f0e114fc184e7e27cc9e369c8fd2e653/pkg/sql/opt/cat/utils_test.go#L167)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Do you have any tests where the schema is provided (e.g. `public`) instead of the database name?
</blockquote></details>

Added (though it's not pretty). One day I will clean all this stuff up and only use `TableName` for resolved tables.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 1 of 2 files at r11, 3 of 4 files at r12, 2 of 2 files at r13, 7 of 7 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37194)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @justinj, and @knz)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
